File: algorithms/leetcode/102-binary-tree-level-order-traversal/binary-tree-level-order-traversal.js
const levelOrder = root => {
    const result = [];

    let node = root;

    const scanLevel = (node, level) => {
        if (!node) {
            return;
        }

        if (!result[level]) {
            result[level] = [];
        }

        result[level].push(node.val);

        scanLevel(node.left, level + 1);
        scanLevel(node.right, level + 1);
    };

    scanLevel(node, 0);

    return result;
};

File: algorithms/leetcode/102-binary-tree-level-order-traversal/BinaryTreeLevelOrderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private void helper(TreeNode root, int level, List<List<Integer>> result) {
        if (root == null) {
            return;
        }
        
        if (level + 1 > result.size()) {
            result.add(new ArrayList<>());
        }
        
        result.get(level).add(root.val);
        
        helper(root.left, level + 1, result);
        helper(root.right, level + 1, result);
    }
    
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        
        helper(root, 0, result);
        
        return result;
    }
}
